### PR TITLE
[exporter][batching] Serialized bytes based batching for logs

### PR DESCRIPTION
**Pull request by sfc-gh-sili**
#### Description

This PR implements serialized bytes based batching.

#### Link to tracking issue

https://github.com/open-telemetry/opentelemetry-collector/issues/3262

#### Testing

```
BenchmarkSplittingBasedOnItemCountManySmallLogs-10                    	     409	   2822478 ns/op	 3148718 B/op	   71082 allocs/op
BenchmarkSplittingBasedOnByteSizeManySmallLogs-10                     	     412	   2890594 ns/op	 3156103 B/op	   71242 allocs/op
BenchmarkSplittingBasedOnItemCountManyLogsSlightlyAboveLimit-10       	      38	  27579702 ns/op	43531582 B/op	  700471 allocs/op
BenchmarkSplittingBasedOnByteSizeManyLogsSlightlyAboveLimit-10        	      24	  43067219 ns/op	43639889 B/op	  702927 allocs/op
BenchmarkSplittingBasedOnItemCountHugeLogs-10                         	      40	  29722710 ns/op	41922925 B/op	  690300 allocs/op
BenchmarkSplittingBasedOnByteSizeHugeLogs-10                          	      16	  64321198 ns/op	42107144 B/op	  694144 allocs/op
```

#### Caveat
We book keep both item count and byte size regardless of batching option, which adds an overhead in both cases.

### CODE REVIEW

File: exporter/exporterbatcher/config.go [hunk 1]
```diff
@@ -41,6 +41,9 @@ type MaxSizeConfig struct {
 	// If the batch size exceeds this value, it will be broken up into smaller batches if possible.
 	// Setting this value to zero disables the maximum size limit.
 	MaxSizeItems int `mapstructure:"max_size_items"`
+
+	// MaxSizeItems is the maximum number of the batch serialized size.
```
Comment (ms-jcorley) [inline exporter/exporterbatcher/config.go:45]: Comment has a copy/paste error:  MaxSizeBytes vs MaxSizeItems.  
Also I'd say "bytes" in the comment, i.e. something like:
//  MaxSizeBytes is the maximum serialized size of the batch in bytes

File: exporter/exporterhelper/logs_batch.go [hunk 2]
```diff
@@ -32,16 +33,27 @@ func (req *logsRequest) MergeSplit(_ context.Context, cfg exporterbatcher.MaxSiz
...
 }
 
 func (req *logsRequest) split(cfg exporterbatcher.MaxSizeConfig) ([]Request, error) {
 	var res []Request
-	for req.ItemsCount() > cfg.MaxSizeItems {
-		ld := extractLogs(req.ld, cfg.MaxSizeItems)
-		size := ld.LogRecordCount()
-		req.setCachedItemsCount(req.ItemsCount() - size)
-		res = append(res, &logsRequest{ld: ld, pusher: req.pusher, cachedItemsCount: size})
+	if cfg.MaxSizeItems != 0 {
```
Comment (ms-jcorley) [inline exporter/exporterhelper/logs_batch.go:43]: If both MaxSizeItems and MaxSizeBytes are set, doesn't this logic only handle MaxSizeItems due to the "else" here?

File: exporter/exporterhelper/logs_batch.go [hunk 3]
```diff
@@ -109,3 +121,99 @@ func resourceLogsCount(rl plog.ResourceLogs) int {
 	}
 	return count
 }
+
+func (req *logsRequest) splitBasedOnByteSize(maxByteSize int) ([]Request, error) {
```
Comment (ms-jcorley) [inline exporter/exporterhelper/logs_batch.go:125]: Is this function unused?

File: exporter/exporterhelper/logs_batch_test.go [hunk 4]
```diff
@@ -123,7 +123,106 @@ func TestMergeSplitLogs(t *testing.T) {
...
 			assert.Equal(t, len(tt.expected), len(res))
 			for i := range res {
-				assert.Equal(t, tt.expected[i], res[i])
+				assert.Equal(t, tt.expected[i].(*logsRequest).ld, res[i].(*logsRequest).ld)
+			}
+		})
+	}
+}
+
+func TestMergeSplitLogsBasedOnByteSize(t *testing.T) {
```
Comment (ms-jcorley) [inline exporter/exporterhelper/logs_batch_test.go:132]: You may want a test or two where both MaxSizeBytes and MaxSizeItems is set.

Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch_test.go:132]: Thank you, Jeremy. I needed some placeholder config API to unblock the implementation, but this config API will be quite different from how it ends up being.

I just opened up https://github.com/open-telemetry/opentelemetry-collector/issues/12303 following some discussion in thishttps://github.com/open-telemetry/opentelemetry-collector/pull/12154. Mainly, we want to
1. Enforces the same sizer for max limit and min limit.
2. Ensure a unified config API for both queuing and batching.

I am happy to hear some feedback and continue discussion there. 

For this PR per se, I wanted the minimal change as possible to unblock implementation, because whichever option we settle on, we will need some refactoring before we can actually read the configuration.

File: exporter/exporterhelper/logs_batch.go (hunk 3 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch.go]: Thanks for catching!

File: exporter/exporterhelper/logs_batch.go (hunk 2 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch.go:43]: See [the other comment](https://github.com/open-telemetry/opentelemetry-collector/pull/12299#discussion_r1943708980)

File: exporter/exporterbatcher/config.go (hunk 1 above)
Comment (sfc-gh-sili) [reply exporter/exporterbatcher/config.go]: Thanks! Updated the comment.

File: exporter/exporterhelper/logs_batch_test.go (hunk 4 above)
Comment (ms-jcorley) [reply exporter/exporterhelper/logs_batch_test.go:132]: I may not be following... why are we adding code for MaxSizeBytes if users can't use it yet?
Looking that the linked issue, wouldn't it be better to wait for the input config format to be determined and do this together as a feature rather than add dead code that will likely need to be changed later?

Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch_test.go:132]: @ms-jcorley Once the batch config API settles, we need to update `MergeSplit` API to support passing in bytes configuration. Right now `MergeSplit` is referenced from 13 files, and 44 lines of code, and all those will need to be updated accordingly.

The implementation is separate though. Whichever API we converge to, the underlying implementation will be same.  I just need the bare minimum so I can invoke the feature from unit test - I picked `MaxSizeBytes` because it requires minimum change.

Comment (ms-jcorley) [reply exporter/exporterhelper/logs_batch_test.go:132]: Ah, that makes more sense to me now.

File: exporter/exporterhelper/logs_batch.go (hunk 2 above)
Comment (ms-jcorley) [reply exporter/exporterhelper/logs_batch.go:43]: Call me paranoid, but since we only expect one to be set, should we fail if we find both?

Comment (dmitryax) [reply exporter/exporterhelper/logs_batch.go:43]: We currently fail if `MaxSizeItems` is [set at all](https://github.com/open-telemetry/opentelemetry-collector/pull/12299/files#diff-2569611ab8d6f62dce2cace7cba426b2452366d2314736ad8fba9231461f5698R62). So we can resolve this comment

File: .chloggen/exporter-batcher-bytes-based-batching.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: exporterhelper
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Support bytes based baching for logs.
```
Comment (dmitryax) [inline .chloggen/exporter-batcher-bytes-based-batching.yaml:10]: I don't think we need a changelog here since it's unavailable for the end user

File: exporter/exporterhelper/logs_batch.go [hunk 6]
```diff
@@ -109,3 +121,87 @@ func resourceLogsCount(rl plog.ResourceLogs) int {
 	}
 	return count
 }
+
+// extractLogs extracts logs from the input logs and returns a new logs with the specified number of log records.
+func extractLogsBasedOnByteSize(srcLogs plog.Logs, capacity int) plog.Logs {
```
Comment (dmitryax) [inline exporter/exporterhelper/logs_batch.go:126]: Have you looked into ways to reuse the existing extract functions by sending a sizer function as an additional argument? It'd be great if we could avoid the code duplication, given that we will need to add other ways to size/split the data later.

File: .chloggen/exporter-batcher-bytes-based-batching.yaml (hunk 5 above)
Comment (sfc-gh-sili) [reply .chloggen/exporter-batcher-bytes-based-batching.yaml:10]: Makes sense. Thanks!

File: exporter/exporterhelper/logs_batch.go (hunk 6 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch.go]: You are right. I abstracted the sizing into an interface called `logSizer` that provides a list of sizing functions.  That way `logSizer` would corresponds to a sizer defined in https://github.com/open-telemetry/opentelemetry-collector/pull/12154. Let me know how it looks.

File: exporter/exporterhelper/logs_batch.go [hunk 7]
```diff
@@ -8,104 +8,127 @@ import (
...
 }
 
 // extractResourceLogs extracts resource logs and returns a new resource logs with the specified number of log records.
-func extractResourceLogs(srcRL plog.ResourceLogs, count int) plog.ResourceLogs {
+func extractResourceLogs(srcRL plog.ResourceLogs, capacity int, sizer sizer.LogsSizer) (plog.ResourceLogs, bool) {
+	capacityReached := false
 	destRL := plog.NewResourceLogs()
 	destRL.SetSchemaUrl(srcRL.SchemaUrl())
 	srcRL.Resource().CopyTo(destRL.Resource())
+	capacityLeft := capacity - sizer.ResourceLogsSize(destRL)
```
Comment (dmitryax) [inline exporter/exporterhelper/logs_batch.go:98]: Is it expected that we don't use `sizer.DeltaSize` here?

File: exporter/exporterhelper/logs_batch.go (hunk 6 above)
Comment (dmitryax) [reply exporter/exporterhelper/logs_batch.go]: It looks good to me. Thanks  you!

File: exporter/exporterhelper/logs_batch.go (hunk 7 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/logs_batch.go:98]: This is intended because this is the first time we take proto size off from capacity. Alternatively this could be:

```
destRL := plog.NewResourceLogs()
capacityLeft = capacity - sizer.Size(destRL)
destRL.SetSchemaUrl(schemaUrl)
capacityLeft -= deltaCapacity(len(schemaUrl)
```

Comment (dmitryax) [reply exporter/exporterhelper/logs_batch.go:98]: Got it. Thanks for clarifying 👍

**Review by dmitryax — APPROVED** [review]

File: exporter/exporterhelper/internal/sizer/logs_sizer.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+// DeltaCapacity() returns the delta size of a proto slice when a new item is added.
+// Example:
+//
+//	prevSize := proto1.Size()
+//	proto1.RepeatedField().AppendEmpty() = proto2
+//
+// Then currSize of proto1 can be calculated as
+//
+//	currSize := (prevSize + deltaCapacity(proto2.Size()))
```
Comment (jmacd) [inline exporter/exporterhelper/internal/sizer/logs_sizer.go:35]:
```suggestion
//	currSize := (prevSize + sizer.DeltaSize(proto2.Size()))
```

File: exporter/exporterhelper/internal/sizer/logs_sizer.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type LogsSizer interface {
+	LogsSize(ld plog.Logs) int
+	ResourceLogsSize(rl plog.ResourceLogs) int
+	ScopeLogsSize(sl plog.ScopeLogs) int
+	LogRecordSize(lr plog.LogRecord) int
+
+	// DeltaCapacity() returns the delta size when a ResourceLog, ScopeLog or LogRecord is added.
```
Comment (jmacd) [inline exporter/exporterhelper/internal/sizer/logs_sizer.go:18]:
```suggestions
	// DeltaSize() returns the delta size when a ResourceLog, ScopeLog or LogRecord is added.
```

File: exporter/exporterhelper/internal/sizer/logs_sizer.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+// Example:
+//
+//	prevSize := proto1.Size()
+//	proto1.RepeatedField().AppendEmpty() = proto2
+//
+// Then currSize of proto1 can be calculated as
+//
+//	currSize := (prevSize + deltaCapacity(proto2.Size()))
+//
+// This is derived from gogo/protobuf 's Size() function
```
[the remark this answers is not in the document]
Comment (jmacd) [reply exporter/exporterhelper/internal/sizer/logs_sizer.go:37]: Yes, and you might add that the constant `1` below is specialized for OTLP; it relies on protobuf tags < 16 (which are expressed w/ a single byte of overhead).

**Review by jmacd — APPROVED** [review]

**Comment by dmitryax** [comment]
@sfc-gh-sili please fix failures in `lint` and `checks` CI jobs
